File: ios/Sources/DKMadsSSPSDK/UI/DKMadsVideoAdController.swift
import Foundation
import UIKit
import AVFoundation

@objc public protocol DKMadsVideoAdControllerDelegate: AnyObject {
    @objc optional func videoAdControllerDidLoad(_ controller: DKMadsVideoAdController)
    @objc optional func videoAdController(_ controller: DKMadsVideoAdController, didFailWithError error: Error)
    @objc optional func videoAdController(_ controller: DKMadsVideoAdController, didEmitEvent eventName: String, payload: [String: Any])
}

/// Loads a video ad and attaches quartile / skip / viewability telemetry to your `AVPlayer`.
@objc public final class DKMadsVideoAdController: NSObject {
    @objc public weak var delegate: DKMadsVideoAdControllerDelegate?
    @objc public let adUnitID: String
    @objc public private(set) var responseInfo: DKMadsResponseInfo?
    @objc public private(set) var loadedAd: Ad?
    @objc public private(set) var isAttached: Bool = false

    public init(adUnitID: String) {
        self.adUnitID = adUnitID
    }

    public func load(
        size: CGSize = CGSize(width: 640, height: 360),
        request: DKMadsAdRequest? = nil,
        completion: ((Result<AdResponse, Error>) -> Void)? = nil
    ) {
        detach()
        SSPSDK.shared.loadAd(
            code: adUnitID,
            format: .video,
            sizes: [size],
            placementCode: request?.placementCode,
            placementContext: request?.placementContext,
            keyValues: request?.keyValues ?? [:]
        ) { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let response):
                self.responseInfo = response.responseInfo
                if response.success, let ad = response.ad {
                    self.loadedAd = ad
                    self.delegate?.videoAdControllerDidLoad?(self)
                } else {
                    let err = NSError(
                        domain: "DKMadsSSPSDK",
                        code: 204,
                        userInfo: [NSLocalizedDescriptionKey: response.reason ?? "no_fill"]
                    )
                    self.delegate?.videoAdController?(self, didFailWithError: err)
                }
                completion?(result)
            case .failure(let error):
                self.delegate?.videoAdController?(self, didFailWithError: error)
                completion?(.failure(error))
            }
        }
    }

    public func attach(
        player: AVPlayer,
        containerView: UIView,
        skippable: Bool? = nil
    ) {
        detach()
        isAttached = true
        let creativeId = loadedAd?.id
        SSPSDK.shared.trackVideoLifecycle(
            adUnitId: adUnitID,
            campaignId: nil,
            creativeId: creativeId,
            player: player,
            containerView: containerView,
            skippable: skippable
        ) { [weak self] event, payload in
            guard let self else { return }
            self.delegate?.videoAdController?(self, didEmitEvent: event, payload: payload)
        }
    }

    @objc public func detach() {
        if isAttached {
            SSPSDK.shared.stopVideoLifecycleTracking(adUnitId: adUnitID)
            isAttached = false
        }
    }

    deinit {
        detach()
    }
}

File: android/DKMadsVideoAdController.kt
package com.dkmads.ssp

import android.net.Uri
import android.view.View
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Loads a video ad and attaches quartile / skip / viewability telemetry via [SSPSDK.trackVideoLifecycle].
 */
class DKMadsVideoAdController(
    val adUnitId: String,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Main + SupervisorJob()),
) {
    interface Listener {
        fun onAdLoaded(ad: Ad, responseInfo: DKMadsResponseInfo) {}
        fun onAdFailed(message: String, responseInfo: DKMadsResponseInfo?) {}
        fun onVideoEvent(eventName: String, payload: Map<String, Any?>) {}
    }

    var listener: Listener? = null
    var loadedAd: Ad? = null
        private set
    var responseInfo: DKMadsResponseInfo? = null
        private set

    private var tracker: VideoTracker? = null
    private var isLoading = false
    private var loadGeneration = 0L

    fun load(
        context: android.content.Context,
        width: Int = 640,
        height: Int = 360,
        placementCode: String? = null,
        placementContext: String? = null,
        keyValues: Map<String, Any> = emptyMap(),
        sizes: List<Pair<Int, Int>>? = null,
    ) {
        if (isLoading) return
        val generation = ++loadGeneration
        isLoading = true
        detach()
        val bidSizes = sizes?.takeIf { it.isNotEmpty() } ?: listOf(width to height)
        scope.launch {
            val result = SSPSDK.loadAd(
                context = context,
                adUnitCode = adUnitId,
                format = AdFormat.VIDEO,
                sizes = bidSizes,
                placementCode = placementCode,
                placementContext = placementContext,
                keyValues = keyValues,
            )
            if (generation != loadGeneration) return@launch
            isLoading = false
            result.fold(
                onSuccess = { ad ->
                    val info = DKMadsResponseInfo.from(ad)
                    responseInfo = info
                    if (!ad.hasFill) {
                        listener?.onAdFailed(ad.reason ?: "no_fill", info)
                        return@fold
                    }
                    if (!ad.isVideo || (ad.playableVideoUrl.isNullOrBlank() && ad.adm.isBlank())) {
                        listener?.onAdFailed("Video fill missing video_url or adm", info)
                        return@fold
                    }
                    loadedAd = ad
                    listener?.onAdLoaded(ad, info)
                },
                onFailure = { err ->
                    listener?.onAdFailed(err.message ?: "load failed", null)
                },
            )
        }
    }

    /** HTTPS MP4 / HLS URL for ExoPlayer or custom players (parity with iOS `playableVideoURL`). */
    fun playbackUri(): Uri? = loadedAd?.playableVideoUrl?.let { Uri.parse(it) }

    fun attach(
        containerView: View,
        durationMsProvider: () -> Long,
        currentPositionMsProvider: () -> Long,
        isPlayingProvider: () -> Boolean,
        isMutedProvider: () -> Boolean = { false },
        skippable: Boolean? = null,
    ) {
        detach()
        tracker = SSPSDK.trackVideoLifecycle(
            adUnitId = adUnitId,
            campaignId = loadedAd?.campaignId,
            creativeId = loadedAd?.creativeId ?: loadedAd?.id,
            containerView = containerView,
            durationMsProvider = durationMsProvider,
            currentPositionMsProvider = currentPositionMsProvider,
            isPlayingProvider = isPlayingProvider,
            isMutedProvider = isMutedProvider,
            skippable = skippable,
            eventListener = { event, payload ->
                listener?.onVideoEvent(event, payload)
            },
        )
    }

    fun detach() {
        tracker?.stop()
        tracker = null
        SSPSDK.stopVideoLifecycleTracking(adUnitId)
    }

    fun destroy() {
        ++loadGeneration
        isLoading = false
        detach()
        scope.cancel()
        loadedAd = null
        responseInfo = null
    }
}
